README.md
<h1 align="center">
  rna-tools
</h1>

<p align="center" style="font-size:20px">
  <b >a toolbox to analyze sequences, structures and simulations of RNA (and way more!)</b>
</p>

<p align="center">
  Look for other our projects at https://github.com/RNA-Puzzles.
</p>


<p>
<div align="center">
	🔥 rna-tools goes online -> https://rna-tools.online 🔥</br></br>
	
<a href='https://ko-fi.com/R5R6AERGM' target='_blank'><img height='36' style='border:0px;height:36px;' src='https://cdn.ko-fi.com/cdn/kofi2.png?v=3' border='0' alt='Buy Me a Coffee at ko-fi.com' /></a>
	
</div>

<p align="center">
  <a href="https://twitter.com/rna_tools"><img src="http://img.shields.io/twitter/follow/rna_tools.svg?style=social&label=Follow"></a>
<!--
  <a href="https://github.com/mmagnus/rna-tools/releases"><img src="https://img.shields.io/github/release/mmagnus/rna-tools.svg"></a>-->
  <a href="https://travis-ci.org/mmagnus/rna-tools"><img src="https://travis-ci.org/mmagnus/rna-tools.svg?branch=master"></a>
  <a href="http://rna-tools.readthedocs.io/en/latest/?badge=latest"><img src="https://readthedocs.org/projects/rna-tools/badge/?version=latest"></a> <a href="https://pypi.org/project/rna-tools"><img src="https://badge.fury.io/py/rna-tools.svg" alt="PyPI version"> </a><a href="https://scrutinizer-ci.com/g/mmagnus/rna-tools/?branch=master"><img src="https://scrutinizer-ci.com/g/mmagnus/rna-tools/badges/quality-score.png?b=master"></a>
	<span class="badge-paypal"><a href="https://www.paypal.me/MarcinMagnus" title="Donate to this project using Paypal"><img src="https://img.shields.io/badge/paypal-donate-1?logo=paypal&color=blue.svg" alt="PayPal donate button" /></a></span>
	
Introduction
-------------------------------------------------------------------------------

<!--
*If you find the tools helpful you can by me a beer via PayPal or Flattr, and if you like the project, please "Star it", so it would be easier to find it for others and to make me happy that the toolbox useful not only for me.*
![](docs/pngs/starit.png)
-->

rna-tools is a core library and a set of programs to run various Python functions related to work, initially, with PDB files of RNA structures, but right now this is a huge toolbox of tools to process various types of RNA data.

**That is why in 2019, after publishing our U6 Molecular Cell paper I decided to rename the package to rna-tools. Simply, various tools to work with RNA data: sequences, alignments, structures, trajectories, RNA-seq data.** If you want access the old version see the [branch](https://github.com/mmagnus/rna-tools/tree/rna-pdb-tools).

The software is used by me in my servers **NPDock** (RNA/DNA-protein docking method, http://genesilico.pl/NPDock/) and **SimRNAweb** (RNA 3D structure prediction method, http://iimcb.genesilico.pl/SimRNAweb/) and **mqapRNA** (RNA 3D quality control, http://iimcb.genesilico.pl/mqapRNA/) and other projects [EvoClustRNA](https://github.com/mmagnus/EvoClustRNA) and [RNA-Puzzles-Normalized-submissions](https://github.com/mmagnus/RNA-Puzzles-Normalized-submissions).

**What is fun here?**

`rna-tools` (formerly rna-pdb-tools) is a packages of shell utils that are using the common core library. You can also access functions of the library from your scripts.

A command-line tools:

```shell
$ rna_pdb_tools.py --is-pdb input/1I9V_A.pdb
True
$ rna_pdb_tools.py --is-pdb input/image.png
False
```

or from a script:

```python
>>> from rna_tools_lib import *
>>> s = RNAStructure('input/1I9V_A.pdb')
>>> s.is_pdb()
True
```

or from a Jupyter Notebook:

<p align="center">
  <img align="center" src="https://raw.githubusercontent.com/mmagnus/rna-tools/master/docs/pngs/align.png")
</p>

Fig. Fetch an alignment and generate an RChie plot for it. See more <https://github.com/mmagnus/rna-tools/blob/master/rna_tools/tools/rna_alignment/rna_alignment.ipynb>

Take a tour [http://mmagnus.github.io/rna-tools/#/](http://mmagnus.github.io/rna-tools/) and/or read the doc [rna-tools.rtfd.io/en/latest/](http://rna-tools.rtfd.io/en/latest/).

<p align="center">
  <img align="center" src="https://raw.githubusercontent.com/mmagnus/rna-tools/master/docs/pngs/qKPVoPxDmq.gif">
</p>

Fig. `rna_pdb_tools.py --get-rnapuzzle-ready *pdb --inplace`

The latest
-------------------------------------------------------------------------------
(see [CHANGELOG](https://github.com/mmagnus/rna-tools/tree/master/CHANGELOG.md) for more detailed description)

**22-05-17** 🔥 a paper published on our server 🔥  
	
rna-tools.online: a Swiss army knife for RNA 3D structure modeling workflow   
Marcin Magnus   
Nucleic Acids Research     
https://doi.org/10.1093/nar/gkac372
	
**22-03-31** rna-tools goes online, http://rna-tools.online
	
**21-05-24** spotifier branched out into own repository to keep RT light https://github.com/mmagnus/yeast-spotifier
	
**20-10/12** [mqapRNA](https://github.com/mmagnus/rna-tools/blob/master/notes/mq.ipynb): py3 wrappers and include them in RT: RASP, Dfire, RNA3DCNN, QRNA, FARNA(Rosetta), AnalyzeGeometry, ClashScore, eSCORE (barnaba), 3dRNAscore, RNAkb (5pt)
    
**20-08-20** A new structures of splicesome processed with rna-tools to be easily viewed with PyMOL (or as single chains) [PyMOL4Spliceosome](https://github.com/mmagnus/PyMOL4Spliceosome)

**20-06-18** A new tool, [spotifier](https://github.com/mmagnus/yeast-spotifier), to process yeast plate images into figures 

**20-03-21** [PyMOL Preview Generator](https://github.com/mmagnus/rna-tools/tree/master/rna_tools/tools/pymol_preview_generator) a new tool for generation of previews in Finders created :-)

**19-11-08** The RNA-Puzzles toolkit paper has been accepted for publication in Nucleic Acid Research :-) Release v3

RNA-Puzzles toolkit: A computational resource of RNA 3D structure benchmark datasets, structure manipulation, and evaluation tools
Magnus, Marcin; Antczak, Maciej; Zok, Tomasz; Wiedemann, Jakub ; Lukasiak, Piotr; Cao, Yang ; Bujnicki, Janusz; Westhof, Eric; Szachniuk, Marta; Miao, Zhichao
https://academic.oup.com/nar/advance-article/doi/10.1093/nar/gkz1108/5651330

**19-10-22** We made a searchable [index]( https://github.com/mmagnus/rna-tools/blob/master/rna-tools-index.csv) of all the tools. There are around 100 functionalities implemented, enjoy it! Let us know if something is missing or unclear!

**19-10-10** **rna-tools finally works with Python 3, to get Python 2 version go to this [branch](https://github.com/mmagnus/rna-tools/tree/py2)** However, not all tools can be used with Python 3, for example, ClaRNA is written in Python 2 and we can do nothing about it. So, for now, we suggest using Conda or something else that supports kind of hybrid Python2/Python3 environments. Read more on this [here](https://towardsdatascience.com/environment-management-with-conda-python-2-3-b9961a8a5097) and [here](https://docs.anaconda.com/anaconda/user-guide/tasks/switch-environment/)

**19-10-01** The EvoClustRNA manuscript with the heavy use of rna-tools is accepted for publication!

M. Magnus, M., Kappel, K., Das, R., & Bujnicki, J. M. (2019). RNA 3D structure prediction guided by independent folding of homologous sequences, BMC Bioinformatics, https://github.com/mmagnus/EvoClustRNA https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-3120-y

**19-06-15** rna-tools used for spliceosome! :-) is accepted for publication! See [this folder](https://github.com/mmagnus/rna-tools/tree/master/U6MolCell) for the description of the analysis!

Eysmont, K., Matylla-Kulinska, K., Jaskulska, A., Magnus, M., & Konarska, M. M. (2019). Rearrangements within the U6 snRNA core at the transition between the two catalytic steps of splicing. Molecular Cell https://github.com/mmagnus/rna-tools/tree/master/U6MolCell https://www.cell.com/molecular-cell/pdfExtended/S1097-2765(19)30390-9 

See also [CHANGELOG](https://github.com/mmagnus/rna-tools/blob/master/CHANGELOG.md).
	
Table of Contents
-------------------------------------------------------------------------------

   * [Tour](#tour)
   * [rna_pdb_tools.py](#rna_pdb_toolspy)
   * [Tools](#tools)
   * [Docs](#docs)
   * [Cite](#cite)
   * [Used in papers](#used-in-papers)
   * [RNA Puzzle Submission](#rna-puzzle-submission)
   * [Inspiration (and alternatives)](#inspiration-and-alternatives)
   * [Install](#install)
   * [Index of Tools](#index-of-tools)
      * [rna\_pdb\_tools\.py](#rna_pdb_toolspy)
      * [Sequence analysis](#sequence-analysis)
      * [Secondary structure analysis](#secondary-structure-analysis)
      * [Tertiary structure comparison](#tertiary-structure-comparison)
      * [Tertiary structure formats](#tertiary-structure-formats)
      * [Tertiary structure analysis](#tertiary-structure-analysis)
      * [Tertiary structure processing](#tertiary-structure-processing)
      * [PyMOL4RNA](#pymol4rna)
      * [SimRNA](#simrna)
      * [Rosetta](#rosetta)
      * [RNA Alignment](#rna-alignment)
      * [Python Classes](#python-classes)
      * [Other](#other)
   * [Index of Jupyter Notebooks](#index-of-jupyters)   
   * [References](#references)

## Tour

Take a tour http://mmagnus.github.io/rna-tools/#/

## rna_pdb_tools.py

```
usage: rna_pdb_tools.py [-h] [--version] [-r] [--no-progress-bar]
                        [--renum-atoms] [--renum-nmr] [--renum-residues-dirty]
                        [--undo] [--delete-anisou] [--fix] [--to-mol2]
                        [--split-alt-locations] [-c] [--is-pdb] [--is-nmr]
                        [--nmr-dir NMR_DIR] [--un-nmr] [--orgmode]
                        [--get-chain GET_CHAIN] [--fetch] [--fetch-ba]
                        [--fetch-chain] [--get-seq] [--color-seq]
                        [--ignore-files IGNORE_FILES] [--compact]
                        [--hide-warnings] [--get-ss] [--rosetta2generic]
                        [--no-hr] [--renumber-residues] [--dont-rename-chains]
                        [--dont-fix-missing-atoms] [--inspect]
                        [--collapsed-view] [--cv] [-v] [--mutate MUTATE]
                        [--edit EDIT] [--rename-chain RENAME_CHAIN]
                        [--swap-chains SWAP_CHAINS] [--set-chain SET_CHAIN]
                        [--replace-chain REPLACE_CHAIN] [--delete DELETE]
                        [--extract EXTRACT] [--extract-chain EXTRACT_CHAIN]
                        [--uniq UNIQ] [--chain-first] [--oneline]
                        [--replace-htm] [--fasta] [--cif2pdb] [--pdb2cif]
                        [--mdr] [--get-rnapuzzle-ready] [--rpr]
                        [--keep-hetatm] [--inplace] [--here] [--suffix SUFFIX]
                        [--replace-hetatm] [--dont-report-missing-atoms]
                        [--backbone-only] [--no-backbone] [--bases-only]
                        file [file ...]

rna_pdb_toolsx - a swiss army knife to manipulation of RNA pdb structures

Usage::

   $ rna_pdb_tools.py --delete A:46-56 --inplace *.pdb

    $ rna_pdb_tools.py --get-seq *
    # BujnickiLab_RNApuzzle14_n01bound
    > A:1-61
    # BujnickiLab_RNApuzzle14_n02bound
    > A:1-61
    CGUUAGCCCAGGAAACUGGGCGGAAGUAAGGCCCAUUGCACUCCGGGCCUGAAGCAACGCG
    [...]

-v is for verbose, --version for version ;)

positional arguments:
  file                  file

optional arguments:
  -h, --help            show this help message and exit
  --version
  -r, --report          get report
  --no-progress-bar     for --no-progress-bar for --rpr
  --renum-atoms         renumber atoms, tested with --get-seq
  --renum-nmr
  --renum-residues-dirty
  --undo                undo operation of action done --inplace, , rename "backup files" .pdb~ to pdb, ALL files in the folder, not only ~ related to the last action (that you might want to revert, so be careful)
  --delete-anisou       remove files with ANISOU records, works with --inplace
  --fix                 fix a PDB file, ! external program, pdbfixer used to fix missing atoms
  --to-mol2             fix a PDB file, ! external program, pdbfixer used to fix missing atoms
  --split-alt-locations
                        @todo
  -c, --clean           get clean structure
  --is-pdb              check if a file is in the pdb format
  --is-nmr              check if a file is NMR-style multiple model pdb
  --nmr-dir NMR_DIR     make NMR-style multiple model pdb file from a set of files 
                        
                          rna_pdb_tools.py --nmr-dir . 'cwc15_u5_fragments*.pdb' > ~/Desktop/cwc15-u5.pdb
                        
                        please use '' for pattern file recognition, this is a hack to deal with folders with
                        thousands of models, if you used only *.pdb then the terminal will complain that you
                        selected to many files.
  --un-nmr              split NMR-style multiple model pdb files into individual models [biopython]
  --orgmode             get a structure in org-mode format <sick!>
  --get-chain GET_CHAIN
                        get chain, one or many, e.g, A, but now also ABC works
  --fetch               fetch file from the PDB db, e.g., 1xjr,
                        use 'rp' to fetchthe RNA-Puzzles standardized_dataset [around 100 MB]
  --fetch-ba            fetch biological assembly from the PDB db
  --fetch-chain         fetch a structure in extract chain, e.g. 6bk8 H
  --get-seq             get seq
  --color-seq           color seq, works with --get-seq
  --ignore-files IGNORE_FILES
                        files to be ingored, .e.g, 'solution'
  --compact             with --get-seq, get it in compact view'
                        $ rna_pdb_tools.py --get-seq --compact *.pdb
                        # 20_Bujnicki_1
                        ACCCGCAAGGCCGACGGCGCCGCCGCUGGUGCAAGUCCAGCCACGCUUCGGCGUGGGCGCUCAUGGGU # A:1-68
                        # 20_Bujnicki_2
                        ACCCGCAAGGCCGACGGCGCCGCCGCUGGUGCAAGUCCAGCCACGCUUCGGCGUGGGCGCUCAUGGGU # A:1-68
                        # 20_Bujnicki_3
                        ACCCGCAAGGCCGACGGCGCCGCCGCUGGUGCAAGUCCAGCCACGCUUCGGCGUGGGCGCUCAUGGGU # A:1-68
                        # 20_Bujnicki_4
                        
  --hide-warnings       hide warnings, works with --get-chain, it hides warnings that given changes are not detected in a PDB file
  --get-ss              get secondary structure
  --rosetta2generic     convert ROSETTA-like format to a generic pdb
  --no-hr               do not insert the header into files
  --renumber-residues   by defult is false
  --dont-rename-chains  used only with --get-rnapuzzle-ready.
                        By default:
                           --get-rnapuzzle-ready rename chains from ABC.. to stop behavior switch on this option
  --dont-fix-missing-atoms
                        used only with --get-rnapuzzle-ready
  --inspect             inspect missing atoms (technically decorator to --get-rnapuzzle-ready without actually doing anything but giving a report on problems)
  --collapsed-view
  --cv                  alias to collapsed_view
  -v, --verbose         tell me more what you're doing, please!
  --mutate MUTATE       mutate residues,
                        e.g.,
                              --mutate "A:1a+2a+3a+4a,B:1a"
                        to mutate to adenines the first four nucleotides of the chain A
                        and the first nucleotide of the chain B
  --edit EDIT           edit 'A:6>B:200', 'A:2-7>B:2-7'
  --rename-chain RENAME_CHAIN
                        edit 'A>B' to rename chain A to chain B
  --swap-chains SWAP_CHAINS
                        B>A, rename A to _, then B to A, then _ to B
  --set-chain SET_CHAIN
                        set chain for all ATOM lines and TER (quite brutal function)
  --replace-chain REPLACE_CHAIN
                        a file PDB name with one chain that will be used to
                        replace the chain in the original PDB file,
                        the chain id in this file has to be the same with the chain id of the original chain
  --delete DELETE       delete the selected fragment, e.g. A:10-16, or for more than one fragment --delete 'A:1-25+30-57'
  --extract EXTRACT     extract the selected fragment, e.g. A:10-16, or for more than one fragment --extract 'A:1-25+30-57'
  --extract-chain EXTRACT_CHAIN
                        extract chain, e.g. A
  --uniq UNIQ           
                        rna_pdb_tools.py --get-seq --uniq '[:5]' --compact --chain-first * | sort
                        A:1-121        ACCUUGCGCAACUGGCGAAUCCUGGGGCUGCCGCCGGCAGUACCC...CA # rp13nc3295_min.out.1
                        A:1-123        ACCUUGCGCGACUGGCGAAUCCUGAAGCUGCUUUGAGCGGCUUCG...AG # rp13cp0016_min.out.1
                        A:1-123        ACCUUGCGCGACUGGCGAAUCCUGAAGCUGCUUUGAGCGGCUUCG...AG # zcp_6537608a_ALL-000001_AA
                        A:1-45 57-71   GGGUCGUGACUGGCGAACAGGUGGGAAACCACCGGGGAGCGACCCGCCGCCCGCCUGGGC # solution
  --chain-first
  --oneline
  --replace-htm
  --fasta               with --get-seq, show sequences in fasta format,
                        can be combined with --compact (mind, chains will be separated with ' ' in one line)
                        
                        $ rna_pdb_tools.py --get-seq --fasta --compact input/20_Bujnicki_1.pdb
                        > 20_Bujnicki_1
                        ACCCGCAAGGCCGACGGC GCCGCCGCUGGUGCAAGUCCAGCCACGCUUCGGCGUGGGCGCUCAUGGGU
                        
  --cif2pdb             [PyMOL Python package required]
  --pdb2cif             [PyMOL Python package required]
  --mdr                 get structures ready for MD (like rpr but without first)

RNAPUZZLE-READY:
  --get-rnapuzzle-ready
                        get RNApuzzle ready (keep only standard atoms).'
                        Be default it does not renumber residues, use --renumber-residues
                        [requires BioPython]
  --rpr                 alias to get_rnapuzzle ready)

CAN BE COMBINED WITH:
  --keep-hetatm         keep hetatoms
  --inplace             in place edit the file! [experimental,
                        only for get_rnapuzzle_ready, --delete, --get-ss, --get-seq, --edit-pdb]
  --here                save a file next to the original file with auto suffix
                        for --extract it's .extr.pdb
  --suffix SUFFIX       when used with --inplace allows you to change a name of a new file, --suffix del will give <file>_del.pdb (mind added _)
  --replace-hetatm      replace 'HETATM' with 'ATOM' [tested only with --get-rnapuzzle-ready]
  --dont-report-missing-atoms
                        used only with --get-rnapuzzle-ready
  --backbone-only       used only with --get-rnapuzzle-ready, keep only backbone (= remove bases)
  --no-backbone         used only with --get-rnapuzzle-ready, remove atoms of backbone (define as P OP1 OP2 O5')
  --bases-only          used only with --get-rnapuzzle-ready, keep only atoms of bases
```

Tricks:

    $ rna_pdb_tools.py --delete A:48-52 --suffix=noloop --inplace
    10_rp17c.out.14.pdb
    10_rp17c.out.14_out.pdb
    [..]

    $ rna_pdb_tools.py --get-rnapuzzle-ready --inplace *.pdb

.. keep original structures in original and use rpr:

    ➜  bujnicki_server_ss for i in original/*.pdb; do rna_pdb_tools.py --get-rnapuzzle-ready $i > ${i/.pdb/_rpr.pdb}; done
    ➜  bujnicki_server_ss ls
    17pz_withSS_all_thrs6.00A_clust01-000001_AA_rpr.pdb 17pz_withSS_all_thrs6.00A_clust06-000001_AA_rpr.pdb
    17pz_withSS_all_thrs6.00A_clust02-000001_AA_rpr.pdb 17pz_withSS_all_thrs6.00A_clust07-000001_AA_rpr.pdb
    17pz_withSS_all_thrs6.00A_clust03-000001_AA_rpr.pdb 17pz_withSS_all_thrs6.00A_clust08-000001_AA_rpr.pdb
    17pz_withSS_all_thrs6.00A_clust04-000001_AA_rpr.pdb 17pz_withSS_all_thrs6.00A_clust09-000001_AA_rpr.pdb
    17pz_withSS_all_thrs6.00A_clust05-000001_AA_rpr.pdb original

.. or to get structures ready for SimRNA/SimRNAweb use :

    $ for i in *pdb; do rna_pdb_tools.py --get-rnapuzzle-ready $i >  ${i/.pdb/_srr.pdb}; done
	# at some point there was a seperate function --get_simrna_ready but there is no need for it
	# simply use --get-rnapuzzle-ready 

rna-tools loves to be used with parallel (wirte in shell script what you want rna-tools to do) and run to execute it in a set of folders:
	
   parallel --bar --eta --progress 'cp test.sh {} && cd {} && bash test.sh ' ::: *
	
## Tools

The (almost) full list of tools can be found here: <https://github.com/mmagnus/rna-tools/blob/master/rna-tools-index.csv>

Read more http://rna-tools.readthedocs.io/en/latest/

## Docs

Read the documentations at [rna-tools.rtfd.io/en/latest/](http://rna-tools.rtfd.io/en/latest/).

<a href="http://rna-tools.rtfd.io/en/latest/"><img src="https://github.com/mmagnus/rna-tools/raw/master/docs/pngs/docs.png"></a>

## Cite

Magnus M, Antczak M, Zok T, Wiedemann J, Lukasiak P, Cao Y, Bujnicki JM, Westhof E, Szachniuk M, Miao Z. RNA-Puzzles toolkit: a computational resource of RNA 3D structure benchmark datasets, structure manipulation, and evaluation tools.  
Nucleic Acids Research. 2019  

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

See https://github.com/mmagnus/rna-tools/releases for more details for 3.7.x

# 3.7

    Fixed:
    
        rna_tools_lib.py (RNAStructure) for remove_hydrogen, correct atom name position now
        rna_pdb_tools.py --get-seq works with protein fix #104	
	    spotifer.py - fix alignments, and lots of other improvements
        
    Added:

        rna_pdb_tools.py:
        - add --fetch-chain to first fetch a structure and extract chain in one step
        
        mqapRNA: py3 wrappers and include them in RT
        - RASP
        - Dfire
        - RNA3DCNN
        - QRNA
        - FARNA
        - AnalyzeGeometry and ClashScore (add rna_clashscore.py app)
        - eSCORE (barnaba)
        - 3dRNAscore
        - RNAkb (!!!!)
        - libs

        mq: add mq to RNAStructure and a simple Jupyter notebook to test it

        rna_tools_lib.py (RNAStructure):
        - add .name attribute with filename (only filename, e.g. "1XJR.pdb")
        - --bases-only and --no-backbone
    
        rna_tools_lib.py: add load_rnas to load structural files (via glob)

        rna_calc_rmsd_all_vs_all.py: add rmsd calc based on pymol
	     Add PyMOL plugin to do that in pymol: PyMOL4RNA.py: rmsdx (get all vs all rmsd matrix)
        
        plotting: add rna_plot_heatmap.py
        
        simrna_trajectory: calc rmsd to another frame

    Changed:
    
        rna_tools_lib.py (RNAStructure) when rpr loads structure with biopython the error is not suppressed
        
# 3.6.2

    Fixed:

        rna_secondary_structure_prediction.py: fixes

    Added:

        PyMOL4RNA.py: mutate, rpr, diff, move rnapuzzle coloring code to own file

# 3.6 201122

    Fixed:

        rna_pdb_tools.py: --rpr fix correct resi for remarks
        rna_pdb_tools.py: --rpr with --keep-hetatm ! [fixed issue with #71]
	
    Added:
    
        rna_pdb_tools.py: --set-chain, set chain for all ATOM lines and TER (quite brutal function)
        rna_pdb_tools.py: add --triple-mode,--column-name
        rna_pdb_tools.py: add --backbone-only
        rna_pdb_tools.py: --rpr no also fixes missing OP1 and OP2 in backbone
        rna_pdb_tools.py: refactor, path code executed once
        rna_pdb_tools.py: add --inspect
   
        rna_tools_lib.py (RNAStructure) - add get_res_text() & get_remarks_text()
        rna_tools_lib.py (RNAStructure) - add reload()
        rna_tools_lib.py (RNAStructure) - change write() now saves by default to self.fn
    
        rna_calc_rmsd_biopython.py: --triple-mode works (final test!) 201118
        rna_calc_rmsd_biopython.py: introduce way to save models after rmsd calc
        rna_calc_rmsd_biopython.py: add --triple-mode,--column-name
		
	PyMOL4RNA: add inspect function for rna_pdb_tools.py: --inspect
		
    Bug fixes and small improvements.

 https://github.com/mmagnus/rna-tools/issues/71

# 3.5.4 200710

- add digitifier.py
- rna_clanstix.py: deal with colors using shapes (type)
- spotifier.py: 
    - clean spotifier.py, -m -> <map>
    - spotifier.py: add logging, process reporters plates, add dont-anntate

# 3.5.3 200703

- spotifier.py: 
    - deal better with fonts
    - add to setup.py spotifier program

# 3.5.2 200702

- spotifier.py: 
    - "--map" should be required=True
    - trim -> dont-align, works on PSD, clean

# 3.5.1 200629

- Fix scrutinizer and small fixes

# 3.5 200629

- rna_calc_inf.py: 
    - --renumber-residues
    - clean sel pdb files, and --dont-remove-sel-files
    - force to run ClaRNA even if <pdb>.outCR file is there, for will be auto True when selection defined
    - rna_calc_inf.py: up readme
    - rna_calc_inf.py: with model_selection and target_selection
    - Fix problems with rna_calc_inf.py, re-write tests, add docs
    
- rna_pdb_tools.py: 
    - add --no-progress-bar
    - add --color-seq
    - add --rpr for --mutate
    
- diffpdb.py: fix --names-and-resi
    
Update update_readme.py and update readme (Used in papers)
    
Bug fixes and small improvements.

# 3.4 200626

200626

- rna_pdb_tools.py: --undo --suffix --hide-warnings
- rna_pdb_tools.py: add fetch RNA-Puzzles standardized_dataset
- rna_pdb_tools.py: --to-mol2
- include mini-moderna3
- add spotifier https://github.com/mmagnus/rna-tools/tree/master/rna_tools/tools/spotifier 
- add webserver-engine https://github.com/mmagnus/rna-tools/tree/master/rna_tools/tools/webserver-engine 

# 3.2

- 200330 read the docs py3 now https://readthedocs.org/projects/rna-tools/builds/
- 190821
    - Add `rna_alignment/rna_align_foldability.py`
    - Seq.py now you can set .name
    - Seq.py mc-fold:
        - Add: explore option
        - Change: verbose now also shows full output of the program
        - Add: comment field for some extra information from the program
- 190820 Add: rna_align_distance_to_seq.py
- 190815 Add: Seq.py: load_fasta_ss_into_RNAseqs()
- 190813 Add: copied and edited from rna_pdb_merge_into_one.py to:

      rna_pdb_tools.py --nmr-dir . 'cwc15_u5_fragments*.pdb' > ~/Desktop/cwc15-u5.pdb

- 190810 Add: ENTRNA wrapper for foldability

## Added

- `rna_dot2ct.py`
- `rna_pdb_tools.py --swap-chains SWAP_CHAINS` [190530]
- Clanstix: with smart group name picking [190500]
- `rna_pdb_tools.py --split-alt-locations`
- `rna_pdb_tools.py --delete-anisou`
- copied and edited from rna_pdb_merge_into_one.py to `rna_pdb_tools.py --nmr-dir . 'cwc15_u5_fragments*.pdb' > ~/Desktop/cwc15-u5.pdb` [190813
]

## Changed
- Rename `rna_ss_pred.py` to `rna_secondary_structure_prediction.py`
- Changed all underscores into dashes in arguments, .e.g --get_seq to --get-seq #94 [190529]
- PyMOL4RNA: scale down shapes for inorganic, to 0.25

## Fixed
- Clanstix problem with group and coloring bug #91

# History

170814 Python3 everywhere (at least it should be)


rna-pdb-tools.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import argparse
import os
import time

from pdb_parser_lib import *

def add_header():
    now = time.strftime("%c")
    print 'HEADER Generated with rna-pdb-tools'
    print 'HEADER ver %s \nHEADER https://github.com/mmagnus/rna-pdb-tools \nHEADER %s' % (version, now)

if __name__ == '__main__':
    version = os.path.basename(os.path.dirname(os.path.abspath(__file__))), get_version(__file__)
    version = version[1].strip()
    parser = argparse.ArgumentParser('rna-pdb-tools.py ver: %s' % version)

    parser.add_argument('-r', '--report', help='get report',
                        action='store_true')
    parser.add_argument('-c', '--clean', help='get clean structure',
                        action='store_true')

    parser.add_argument('--get_chain', help='get chain, .e.g A')

    parser.add_argument('--get_seq', help='get seq', action='store_true')

    parser.add_argument('--rosetta2generic', help='convert ROSETTA-like format to a generic pdb',
                        action='store_true')

    parser.add_argument('--get_rnapuzzle_ready', help='get RNApuzzle ready (keep only standard atoms, renumber residues)',
                        action='store_true')

    parser.add_argument('--no_hr', help='do not insert the header into files',
                        action='store_true')

    parser.add_argument('--renumber_residues', help='',
                        action='store_true')

    parser.add_argument('--get_simrna_ready', help='',
                        action='store_true')

    parser.add_argument('--edit',
			dest="edit",
                        default='',
                        help="edit 'A:6>B:200', 'A:2-7>B:2-7'")

    parser.add_argument('--delete',# type="string",
			dest="delete",
			default='',
			help="delete the selected fragment, e.g. A:10-16")
    
    parser.add_argument('file', help='file')
    #parser.add_argument('outfile', help='outfile')   

    args = parser.parse_args()

    s = StrucFile(args.file)
    if args.report:
        print s.get_report()
        print s.get_preview()

    s = StrucFile(args.file)
    if args.clean:
        s.decap_gtp()
        s.fix_resn()
        s.remove_hydrogen()
        s.remove_ion()
        s.remove_water()
        s.renum_atoms()
        s.fix_O_in_UC()
        s.fix_op_atoms()
        #print s.get_preview()
        #s.write(args.outfile)
        if not args.no_hr:
            add_header()
        print s.get_text()

    s = StrucFile(args.file)
    if args.get_seq:
        s.decap_gtp()
        s.fix_resn()
        s.remove_hydrogen()
        s.remove_ion()
        s.remove_water()
        s.renum_atoms()
        s.fix_O_in_UC()
        s.fix_op_atoms()
        #print s.get_preview()
        print s.get_seq()

    s = StrucFile(args.file)
    if args.get_chain:
        s.fix_resn()
        s.remove_hydrogen()
        s.remove_ion()
        s.remove_water()
        s.renum_atoms()
        s.fix_O_in_UC()
        s.fix_op_atoms()
        #print s.get_preview()
        print s.get_chain(args.get_chain)

    if args.rosetta2generic:
        s = StrucFile(args.file)
        s.fix_resn()
        s.remove_hydrogen()
        s.remove_ion()
        s.remove_water()
        s.fix_op_atoms()
        s.renum_atoms()
        #print s.get_preview()
        #s.write(args.outfile)
        if not args.no_hr:
            add_header()
        print s.get_text()

    if args.get_rnapuzzle_ready:
        s = StrucFile(args.file)
        s.decap_gtp()
        s.fix_resn()
        s.remove_hydrogen()
        s.remove_ion()
        s.remove_water()
        s.fix_op_atoms()
        s.renum_atoms()
        #print s.get_preview()
        #s.write(args.outfile)
        if not args.no_hr:
            add_header()
        s.get_simrna_ready(args.renumber_residues)
        print s.get_text()

    if args.get_simrna_ready:
        s = StrucFile(args.file)
        s.decap_gtp()
        s.fix_resn()
        s.remove_hydrogen()
        s.remove_ion()
        s.remove_water()
        s.fix_op_atoms()
        s.renum_atoms()
        if not args.no_hr:
            add_header()
        s.get_simrna_ready(args.renumber_residues)
        print s.get_text()

    if args.delete:
        selection = select_pdb_fragment(args.delete)
        s = StrucFile(args.file)
        if not args.no_hr:
            add_header()
            print 'HEADER --delete ' + args.delete #' '.join(str(selection))
        for l in s.lines:
            if l.startswith('ATOM'):
                chain = l[21]
                resi = int(l[23:26].strip())
                if selection.has_key(chain):
                    if resi in selection[chain]:
                        continue  # print chain, resi
                print l

    if args.edit:
        selection_from, selection_to = select_pdb_fragment(args.edit.split('>')[0]), select_pdb_fragment(args.edit.split('>')[1])
        if len(selection_to) != len(selection_from):
            raise Exception('len(selection_to) != len(selection_from)')
        s = StrucFile(args.file)
        if not args.no_hr:
            add_header()
            print 'HEADER --edit ' + args.edit
        c = 0
        resi_prev = None
        for l in s.lines:
            if l.startswith('ATOM'):
                chain = l[21:22].strip()
                resi = int(l[23:26].strip())
                if selection_from.has_key(chain):
                    if resi in selection_from[chain]:
                        if resi != resi_prev and resi_prev:
                            c += 1
                        resi_prev = resi

                        nl = list(l)
                        nl[21] =  selection_to.keys()[0]
                        nl[23:26] = str(selection_to[selection_to.keys()[0]][c]).rjust(3)
                        nl = ''.join(nl)
                        print nl
                    else:
                        print l

if __name__ == '__main__':
    pass
